### DISCUSSION TITLE
How to interpret insertions with positions in `r.`

### DISCUSSION THREAD

**Discussion by mihailefter** [HGVSnomenclature/hgvs-nomenclature#49] [Q&A]
For the following description `NC_000023.10(NM_004006.2):r.175_176ins180_191` does the inserted sequence include also the transcribed intron (4869 bases long) between `c.186` (`g.32867845`) and `c.187` (`g.32862977`) or just the 12 bases between `r.180` and `r.191` from the actual RNA sequence?

**Comment by ifokkema** [comment]
Since `r.175_176ins180_191` indicates that `r.180_191` is inserted, these are 12 bases. The `r.` prefix doesn't have positions for intronic bases.

**Reply by mihailefter** [reply]
There are however some examples ([`NG_012232.1(NM_004006.2):r.2949_2950ins[2950-30_2950-12;2950-4_2950-1]`](https://github.com/HGVSnomenclature/hgvs-nomenclature/blob/b43f79525913bdaab6990d9ad0ef7e1370fb04a7/docs/recommendations/RNA/insertion.md?plain=1#L34) and [`NC_000023.10(NM_004006.2):r.186_187ins186+1_186+4`](https://github.com/HGVSnomenclature/hgvs-nomenclature/blob/b43f79525913bdaab6990d9ad0ef7e1370fb04a7/docs/background/numbering.md?plain=1#L52)) in which intronic positions are mentioned. Maybe for the second one something like `NC_000023.10(NM_004006.2):r.186_187insguau` or `NC_000023.10(NM_004006.2):r.186_187insNC_000023.10(NM_004006.2):c.186+1_186+4` should be used instead?

**Reply by ifokkema** [reply]
Good catch! Two issues of the current documentation are that rules are sometimes "defined" by examples, and examples are sometimes either missing, inconsistent, or incorrect.
- That the `r.` prefix has no positions for intronic bases is currently not explicitly defined. Having intronic positions for the `r.` prefix, IMO, makes as much sense as having UTR positions for protein descriptions.
- Yet, there are examples given where the `r.` prefix is used to indicate intronic positions.
- A possible solution would be to replace `r.` with `c.`, since an NM reference sequence is given, but there is no definition of whether this is allowed and how this should look. E.g., is `NC_000023.10(NM_004006.2):r.186_187ins[c.186+1_186+4]` a possible variant description? Or should it be `NC_000023.10(NM_004006.2):r.186_187ins[NC_000023.10(NM_004006.2):c.186+1_186+4]`?
- In the examples above, I assumed that using a different prefix requires using square brackets, but this is not defined anywhere. Using a second reference sequence does require square brackets.

I'll see if I can have this added to the next HVNC meeting agenda.

**Reply by mihailefter** [reply]
In this respect maybe the following image would be more helpful than the [current one](https://github.com/HGVSnomenclature/hgvs-nomenclature/blob/b43f79525913bdaab6990d9ad0ef7e1370fb04a7/docs/background/numbering.md?plain=1#L95)?

![coordinate_systems](https://github.com/HGVSnomenclature/hgvs-nomenclature/assets/21262869/226d7b6c-7fa8-473b-9d43-b75c0d6bdb2e)

**Reply by reece** [reply]
Great figure @mihailefter. That would be a very helpful update to the docs.

Would you be open to recreating it in drawio and committing it with the hgvs-nomenclature docs? (vscode has a great plugin for drawio to make this easy)

**Reply by mihailefter** [reply]
@reece recreating it in drawio is not an option. Some automatic conversion from SVG to drawio could be an alternative. But, what's wrong with the SVG format?

**Reply by reece** [reply]
@mihailefter : I didn't realize that that it was in SVG already. That's fine. I'm satisfied as long as it's in an open format that's easily updated by the community.

**Reply by mihailefter** [reply]
@reece: Is it OK then to add it as such? Should it replace the current one?

**Reply by reece** [reply]
@mihailefter Yes, your diagram is a significant improvement over the current. Thank you.

I'll open a new issue and PR for this. I have a few updates to propose that builds on your excellent work.

https://github.com/HGVSnomenclature/hgvs-nomenclature/pull/156 (currently draft)

**Comment by jfjlaros** [comment]
> * A possible solution would be to replace `r.` with `c.`, [...]

Unfortunately, this also means that a sequence of a different molecular type is inserted.

> [...] Using a second reference sequence does require square brackets.

It is not required by the HGVS syntax rules (which status is unclear). Although most examples seem to use brackets, there is also a counterexample: [`r.123_124insL37425.1:r.23_361`](https://github.com/HGVSnomenclature/hgvs-nomenclature/blob/b43f79525913bdaab6990d9ad0ef7e1370fb04a7/docs/recommendations/RNA/insertion.md?plain=1#L24).

**Reply by ifokkema** [reply]
> > * A possible solution would be to replace `r.` with `c.`, [...]
> 
> Unfortunately, this also means that a sequence of a different molecular type is inserted.

Just like when, e.g., an NC-based sequence is inserted between NM-based positions, or when switching between NM and NC(NM) reference sequences.


> > [...] Using a second reference sequence does require square brackets.
> 
> It is not required in the HGVS syntax rules

See:
> "[ ]" (square brackets) are used for alleles (...) and insertions from a second reference sequence

([source](https://hgvs-nomenclature.org/en/latest/recommendations/general/))


>  (which status is unclear).

What is unclear?


> Although most examples seem to use brackets, there is also a counterexample: [`r.123_124insL37425.1:r.23_361`](https://github.com/HGVSnomenclature/hgvs-nomenclature/blob/b43f79525913bdaab6990d9ad0ef7e1370fb04a7/docs/recommendations/RNA/insertion.md?plain=1#L24).

The examples in the docs are chock-full of errors. The page I linked to above has an error where a `g.` prefix is left out after an NC reference sequence, but we all know that the prefix is always required. Mistakes are (so far) always in places where the focus is not on that feature; e.g., where the prefix is explained, it's never missing. But, I found 8 examples where it was missing (I already submitted a PR for this). The square brackets were not missing where their use was explained, but, as you noticed, they are forgotten in a very small number of examples. I will include the omission you mentioned in my last PR.

**Reply by jfjlaros** [reply]
> Just like when, e.g., an NC-based sequence is inserted between NM-based positions, or when switching between NM and NC(NM) reference sequences.

As long as the molecular type is equal, this is not a problem. E.g., I see no problems with the following:

    NM_002001.3:r.100_101insNG_012337.3(NM_003002.4):r.10_15

> What is unclear?

I was referring to the formal HGVS syntax specification, which is used by a number of HGVS parsers. As far as I know this specification not endorsed by the HVNC.

**Reply by jfjlaros** [reply]
Some additional thoughts.

If the second numbering system identifier is defined to be the input for both the molecular type as well as the numbering system (obviously) of the inserted sequence and the first numbering system identifier as the output, insertions of incompatible molecular types are possible via some (reverse) transcription or (back) translation mechanism.

For example,

    NG_012337.3(NM_003002.4):r.100:101insc.52+1_52+3

would be interpreted as: take the DNA sequence between `c.52+1` and `c.52+3`, translate it to RNA and insert it.

This would also allow for insertions of sequences that are not part of any transcript, e.g.,

    NM_002001.2:r.100_101insNG_012337.3:g.10_12

with the associated protein description:

    NM_002001.2:p.Ser33Leu34insNG_012337.3:g.10_12

Additionally, this would allow us to describe inserted reverse transcribed RNA (e.g., as the result of retrovirus activity).

    NG_012337.3:g.10_12insNM_002001.2:r.1_500

---

It is unclear to me whether or not these ideas conflict with the following remark [[1](https://hgvs-nomenclature.org/en/latest/recommendations/DNA/alleles/)]:

> - descriptions combining variants based on different reference sequence types (e.g. c.[76A>C];g.[10091C>G]) should not be used.

**Reply by ifokkema** [reply]
> > Just like when, e.g., an NC-based sequence is inserted between NM-based positions, or when switching between NM and NC(NM) reference sequences.
> 
> As long as the molecular type is equal, this is not a problem.

Whether it's a problem at all doesn't seem to be defined anywhere on the current website, so it'll need to be discussed (see the email that I sent recently).


> > What is unclear?
> 
> I was referring to the formal HGVS syntax specification, which is used by a number of HGVS parsers. As far as I know this specification not endorsed by the HVNC.

It would be good to get it endorsed then, but then we'll first need a full list of differences between this formal specification and the official nomenclature guidelines. That, however, will not be an easy task. We both know that sometimes it's unclear whether something is a rule, a recommendation, or simply an example.


> with the associated protein description:
> 
> ```
> NM_002001.2:p.Ser33Leu34insNG_012337.3:g.10_12
> ```

(You meant to use an NP reference sequence there, right?)
I'm wondering if this will leave too much room for non-unique variant descriptions; if `g.10_12` is normally part of the coding sequence and it normally will be translated in that frame, the same insertion can be given using an NP reference sequence. The question would simply become, why not insert a protein sequence, as that is much simpler?
(obviously, we're talking about only one amino acid in this example, so we'd just name the amino acid, but I assume you meant this as an example for longer insertions)

Also, unless this variant description is a prediction, the question would be how the user would know the source of the inserted protein sequence is, in fact, that piece of genomic DNA. But I guess, also here, the length of the insertion plays a role.


> It is unclear to me whether or not these ideas conflict with the following remark [[1](https://hgvs-nomenclature.org/en/latest/recommendations/DNA/alleles/)]:
> 
> > * descriptions combining variants based on different reference sequence types (e.g. c.[76A>C];g.[10091C>G]) should not be used.

That page is about alleles, so combined variant descriptions that always need to be on the same molecular level. An allele does not exist of, e.g., a DNA and an RNA variant. The examples above were all just one variant description. But we need this clarified, indeed.

**Reply by jfjlaros** [reply]
> I'm wondering if this will leave too much room for non-unique variant descriptions

Descriptions that contain inserted foreign sequences do not have a unique representation in general. E.g., the inserted sequence can be present in multiple reference sequences.

As a rule of thumb, we could say that the molecular type of the inserted sequence should be as close as possible to that of the output.

**Comment by murphyte** [comment]
I believe there's software that will scan full text for HGVS-like strings and submit them for validation, for use with manuscripts? Could that also be used for the web pages, rather than having to go through and validate every expression individually?

**Reply by ifokkema** [reply]
Although this may work, it will work only for a subset of the variant descriptions. The documentation contains lots of examples that current validators don't handle because, on the sequence level, they can not be validated; e.g., variants with unclear boundaries. Also, currently, validators have limited support for RNA and protein descriptions. And finally, @jfjlaros gave an example of where the HGVS syntax rules used in validators differ from what's on the website; e.g., an insertion requiring square brackets when the inserted sequence is described using a reference sequence identifier. It may come down to us going over some of these differences and deciding which is correct; the examples on the website, or the definitions in variant validation software.

**Reply by reece** [reply]
I agree that we could (and should) scrutinize every example in the docs. One way might be to extract anything that looks broadly like a variant (e.g., something as bad as `\w[_\w()]+:[cgmnopr]\.[-_\w;,()]+`), then run unique putative variants through variuos parsers. The goal is mostly to identify broken examples, but I'm sure that it would also identify non-compliant parsers. (The biocommons hgvs Python library  will for sure fail on a bunch of the HGVS Nomenclature examples.) Any example that fails should be scrutinized for correctness.

We also might do well to actually have fewer well-chosen and hand-curated examples. As with code testing, duplicative tests actually obscures problems. Curated examples could be the basis for a compliance suite eventually.

**Reply by jfjlaros** [reply]
The word "doctest" comes to mind.

**Comment by jtdendunnen** [comment]
Based on these discussions, what would be proposal to cover this topic properly?

**Reply by ifokkema** [reply]
During the previous HVNC meeting, we agreed that:
- The mRNA molecule does not have introns, therefore, `r.` variants do not have intronic positions.
- I believe, but I'm not 100% sure anymore, that it was decided that inserting cDNA positions into an RNA reference sequence was acceptable, but it will get more complicated if we would allow other types of "cross-molecule" insertions as well.
- This would mean that intron retention should be described as `NM_004006.2:r.186_187ins[NC_000023.10(NM_004006.2):c.186+1_186+4]` or perhaps `NC_000023.10(NM_004006.2):r.186_187ins[c.186+1_186+4]`.

**Reply by jtdendunnen** [reply]
I agree with a format like NC_000023.10(NM_004006.2):r.186_187ins[c.186+1_186+4]. Sometimes there are insertions from other places in the genome and following this format I assume these can be described as NC_000023.10(NM_004006.2): 	r.94_264delins[NC_000008.10:g.16369356_16369419inv]. Correct?

**Reply by jfjlaros** [reply]
That was the idea, indeed.

**Reply by jtdendunnen** [reply]
I have given the format "NC_000023.10(NM_004006.2):r.186_187ins[c.186+1_186+4]" some more thought but think we should not use c. My reasoning:
1) everybody will understand the format r.186_187ins186+1_186+4, why unnecessarily complicate HGVS nomenclature.
2) the format suggests the inserted sequence (c.186+1_186+4) is "GTAT", which is not correct, inserted is the sequence "guau".
3) indeed the mature RNA sequence (mRNA) does not contain introns but the original messenger RNA does. We could reason the unspliced RNA is the reference sequence.  The recommendations for the RNA reference sequence have: "nucleotide numbering for a RNA reference sequencing follows that of the associated coding or non-coding DNA reference sequence; nucleotide r.123 relates to c.123 or n.123".

**Reply by ifokkema** [reply]
I'm afraid that's simply not possible. If the unspliced RNA is the reference sequence, we are unable to have _any_ variant description on the RNA level showing an effect of splicing.

**Reply by jtdendunnen** [reply]
Agree, see what you mean, but this only applies to my point 3.

**Reply by ifokkema** [reply]
Sorry, I considered them related.
- Considering the first point; it's ambiguous and inconsistent.
  - Ambiguous: would `r.1_2ins100_200` include intronic sequences or not? To have this depend on whether an `NM` or `NC(NM)` projection is used, would be dangerous, in my opinion.
  - Consistency: it would be yet another exception. _"`r.` variants have no introns, except..."_.
- I don't think that the second interpretation is necessarily true. Sure, the NM reference sequence is a DNA reference sequence. But when we use the `r.` numbering, we all agree that we're no longer talking about the DNA sequence but about the RNA sequence. However, that sequence isn't actually included in the reference sequence. The same would hold for using `NC(NM)` projections, anyway necessary for referring to intronic positions. So `NC(NM):r....` already "transcribes" the DNA sequence in that reference sequence to RNA, assuming all `T`s become `u`s, etc. I consider it a good option to use the same "magic" for this variant description.

**Reply by ifokkema** [reply]
Moving the discussion on this part to the [issue](https://github.com/HGVSnomenclature/hgvs-nomenclature/issues/147) on the implementation.

**Comment by reece** [comment]
A doctest is a technique in which an example is pulled from documentation and actually tested. This makes it essentially impossible to include broken examples unintentionally.

It should be fairly easy to extract and validate full HGVS expressions (i.e., with sequence identifiers, etc). It's less clear how we could automatically test fragments like `p.Ala17Trp`, but we could at least reject examples like `c.A>Q`. 

We would implement these as tests that are executed on commit and/or during the documentation build process.

**Reply by ifokkema** [reply]
LOVD's [HGVS syntax library](https://api.lovd.nl/v1/checkHGVS/c.157C%3ET) allows for validation of syntax (no sequence-level validation) of DNA variants without reference sequences. We aim to support all valid HGVS syntax; since no sequence-level validation is performed, this should be possible. We do currently have some issues with some of the repeat syntax, which I noticed recently.
